examples/server/tests/features/server.feature

```diff
@@ -0,0 +1,100 @@
+Feature: llama.cpp server
+
+  Background: Server startup
+    Given a server listening on localhost:8080 with 2 slots and 42 as seed
+    Then  the server is starting
+    Then  the server is healthy
+
+  @llama.cpp
+  Scenario: Health
+    When the server is healthy
+    Then the server is ready
+    And  all slots are idle
+
+  @llama.cpp
+  Scenario Outline: Completion
+    Given a <prompt> completion request with maximum <n_predict> tokens
+    Then  <predicted_n> tokens are predicted
+
+    Examples: Prompts
+      | prompt                           | n_predict | predicted_n |
+      | I believe the meaning of life is | 128       | 128         |
+      | Write a joke about AI            | 512       | 512         |
+
+  @llama.cpp
+  Scenario Outline: OAI Compatibility
+    Given a system prompt <system_prompt>
+    And   a user prompt <user_prompt>
+    And   a model <model>
+    And   <max_tokens> max tokens to predict
+    And   streaming is <enable_streaming>
+    Given an OAI compatible chat completions request
+    Then  <predicted_n> tokens are predicted
+
+    Examples: Prompts
+      | model        | system_prompt               | user_prompt                          | max_tokens | enable_streaming | predicted_n |
+      | llama-2      | You are ChatGPT.            | Say hello.                           | 64         | false            | 64          |
+      | codellama70b | You are a coding assistant. | Write the fibonacci function in c++. | 512        | true             | 512         |
+
+  @llama.cpp
+  Scenario: Multi users
+    Given a prompt:
+      """
+      Write a very long story about AI.
+      """
+    And a prompt:
+      """
+      Write another very long music lyrics.
+      """
+    And 32 max tokens to predict
+    Given concurrent completion requests
+    Then the server is busy
+    And  all slots are busy
+    Then the server is idle
+    And  all slots are idle
+    Then all prompts are predicted
+
+  @llama.cpp
+  Scenario: Multi users OAI Compatibility
+    Given a system prompt "You are an AI assistant."
+    And a model tinyllama-2
+    Given a prompt:
+      """
+      Write a very long story about AI.
+      """
+    And a prompt:
+      """
+      Write another very long music lyrics.
+      """
+    And 32 max tokens to predict
+    And streaming is enabled
+    Given concurrent OAI completions requests
+    Then the server is busy
+    And  all slots are busy
+    Then the server is idle
+    And  all slots are idle
+    Then all prompts are predicted
+
+  # FIXME: #3969 infinite loop on the CI, not locally, if n_prompt * n_predict > kv_size
+  @bug
```

Comment:
@ngxson @ggerganov I am facing a weird scenario where it passed locally but [not on github worker.](https://pipelinesghubeus24.actions.githubusercontent.com/jEJM6zrMhVG8NHAbd8fFWzevGGKuKhmDZUCVPMvRQzymMWYX1b/_apis/pipelines/1/runs/100/signedlogcontent/2?urlExpires=2024-02-20T22%3A20%3A33.3282517Z&urlSigningMethod=HMACV1&urlSignature=jQrLQGljIPhyP4%2FPLVhH41uU8L%2FLtMqoJ4MQQXg0vW4%3D) It looks if the total number of token to generate accross slots exceeds the kv size, the server goes to infinite loop, see #3969.

You can grep for `steps.py` in the workflow output to see that the server is busy forever while the scenario wait for idle status.

You can run locally if you replace `@bug`  by `@llama.cpp` in the feature file.

Reply:
When I run it locally, it fails earlier:

<details>

```
Feature: llama.cpp server # features/server.feature:1

  Background: Server startup  # features/server.feature:3

  @llama.cpp
  Scenario: Health                                                                               # features/server.feature:9
    Given a server listening on localhost:8080 with 2 slots, 42 as seed and llama.cpp as api key # features/steps/steps.py:10 0.000s
    Then the server is starting                                                                  # features/steps/steps.py:27 0.004s
    Then the server is healthy                                                                   # features/steps/steps.py:27
{"timestamp":1708501224,"level":"INFO","function":"log_server_request","line":2398,"message":"request","remote_addr":"127.0.0.1","remote_port":57462,"status":200,"method":"GET","path":"/health","param    Then the server is healthy                                                                   # features/steps/steps.py:27 0.011s
    When the server is healthy                                                                   # features/steps/steps.py:27
{"timestamp":1708501224,"level":"INFO","function":"log_server_request","line":2398,"message":"request","remote_addr":"127.0.0.1","remote_port":57463,"status":200,"method":"GET","path":"/health","param    When the server is healthy                                                                   # features/steps/steps.py:27 0.007s
    Then the server is ready                                                                     # features/steps/steps.py:27
{"timestamp":1708501224,"level":"INFO","function":"log_server_request","line":2398,"message":"request","remote_addr":"127.0.0.1","remote_port":57464,"status":200,"method":"GET","path":"/health","param    Then the server is ready                                                                     # features/steps/steps.py:27 0.005s
      Traceback (most recent call last):
        File "/Users/ggerganov/development/github/llama.cpp/.venv/lib/python3.11/site-packages/behave/model.py", line 1329, in run
          match.run(runner.context)
        File "/Users/ggerganov/development/github/llama.cpp/.venv/lib/python3.11/site-packages/behave/matchers.py", line 98, in run
          self.func(context, *args, **kwargs)
        File "features/steps/steps.py", line 42, in step_wait_for_the_server_to_be_started
          wait_for_health_status(context, 200, 'ok',
        File "features/steps/steps.py", line 324, in wait_for_health_status
          and (slots_idle is None or health['slots_idle'] == slots_idle)
                                     ~~~~~~^^^^^^^^^^^^^^
      KeyError: 'slots_idle'

    And all slots are idle                                                                       # None


Failing scenarios:
  features/server.feature:9  Health

0 features passed, 1 failed, 0 skipped
0 scenarios passed, 1 failed, 0 skipped, 17 untested
4 steps passed, 1 failed, 1 skipped, 0 undefined, 142 untested
Took 0m0.027s
```
</details>

Reply:
Strange I have added those fields for all `200` in #5594. Have you rebased/built with the latest master ?

Reply:
Ah, I didn't notice it was picking the `server` from `build/bin`. It works now

Reply:
Running the problematic test with thread sanitizers surfaces the issue locally:

```bash
cmake -DLLAMA_SANITIZE_THREAD=ON .
```

We should add 2 more CI runs - one with thread sanitizer and one with address sanitizer enabled.

In the meantime, I will try to figure out the root cause for the failure and fix it

Reply:
Running the tests with thread sanitizer enabled we can see a few problems with the current implementation. The endpoints "/health" and "/slots" are incorrectly accessing the slot data creating a data race. Instead, such requests should be forwarded as tasks

Even if this isn't the root cause for #3969, we first need to fix this before proceeding.

It seems to me that the fix might involve a larger refactoring of `server` so it's probably better to first merge this PR (even if the CI fails) and work in a new PR to resolve the issues

Reply:
Thanks for identifying the data race. As I did changes on `/health` and `/slots` endpoint in PR #5550 and #5548. I will submit a patch ASAP to use tasks. I was wrongly thinking as it's just accessing slot data in read only, no race condition will occur.